### DISCUSSION TITLE
Add a way to specify format style on the clangd command line (perhaps using a --style flag like clang-format's)

### DISCUSSION THREAD

**Discussion by leet0rz** [clangd/clangd#1904] [General]
When I run lsp format it forces the indent to 2 spaces, possible to make this 4 spaces in the settings under lspconfig maybe?
Thanks.

**Comment by HighCommander4** [comment]
Clangd respects [clang-format options](https://clang.llvm.org/docs/ClangFormatStyleOptions.html), so creating a `.clang-format` file in the project directory with `IndentWidth: 4` should do it.

**Reply by wangp-blog** [reply]
Do I have to install clang-format in order to format the code? I just installed clangd

**Comment by leet0rz** [comment]
> Clangd respects [clang-format options](https://clang.llvm.org/docs/ClangFormatStyleOptions.html), so creating a `.clang-format` file in the project directory with `IndentWidth: 4` should do it.

Would it be possible to set this under the language server itself as a setting?

**Reply by HighCommander4** [reply]
Could you elaborate on the motivation for this? We generally try to avoid duplicating configuration in multiple places.

**Comment by leet0rz** [comment]
> Could you elaborate on the motivation for this? We generally try to avoid duplicating configuration in multiple places.

It would be cleaner to just have it under lspconfig/language server as a setting for the language server itself as most of the other ones do, most of them respect the on_attach that comes with neovim at least but that does not seem to be the case here. At least in my opinion it would be cleaner than spreading dot files all over the place.

**Reply by HighCommander4** [reply]
Sorry, I'm not a vim user and I'm not familiar with what lspconfig or on_attach are. Is it a way to pass command line flags to the language server executable?

If the request is to allow specifying the format style using a command-line argument to clangd, that might be doable.

`clang-format` has a `--style` argument which is pretty versatile:
 ```
  --style=<string>               - Set coding style. <string> can be:
                                   1. A preset: LLVM, GNU, Google, Chromium, Microsoft,
                                      Mozilla, WebKit.
                                   2. 'file' to load style configuration from a
                                      .clang-format file in one of the parent directories
                                      of the source file (for stdin, see --assume-filename).
                                      If no .clang-format file is found, falls back to
                                      --fallback-style.
                                      --style=file is the default.
                                   3. 'file:<format_file_path>' to explicitly specify
                                      the configuration file.
                                   4. "{key: value, ...}" to set specific parameters, e.g.:
                                      --style="{BasedOnStyle: llvm, IndentWidth: 8}"
```

Clangd could potentially accept the same thing and interpret it the same way, so that you could pass e.g.:

```
--style="{IndentWidth: 4}"
```

**Reply by heygarrett** [reply]
That would be great! If `clangd` had a `--style` flag then something like this would be possible in Neovim to effectively set default style options when a `.clang-format` file isn't present:

```lua
require("lspconfig")["clangd"].setup({
	on_new_config = function(config)
		if vim.tbl_isempty(vim.fs.find(".clang-format", { upward = true })) then
			config.cmd = { "clangd", '--style="{IndentWidth: 4}"' }
		end
	end,
})
```

**Reply by leet0rz** [reply]
> That would be great! If `clangd` had a `--style` flag then something like this would be possible in Neovim to effectively set default style options when a `.clang-format` file isn't present:
> 
> ```lua
> require("lspconfig")["clangd"].setup({
> 	on_new_config = function(config)
> 		if vim.tbl_isempty(vim.fs.find(".clang-format", { upward = true })) then
> 			config.cmd = { "clangd", '--style="{IndentWidth: 4}"' }
> 		end
> 	end,
> })
> ```

I would prefer this over a file in the system. It's just nicer to have this config in the lsp itself rather than in a dot file. Makes the whole setup simpler as it's just in the settings when you clone it to different computers while at the same time not having to create files to keep all over the place.

**Comment by leet0rz** [comment]
```
              lspconfig.clangd.setup({
                  on_attach = custom_attach,
                  capabilities = capabilities,
                  settings = {
                      ccls = {
                          completion = {
                              filterAndSort = false
                          }
                      },
                      clangd = {
                          cmd = { 'clangd', '--background-index', '--header-insertion=iwyu' },
                          filetypes = { 'c', 'cpp', 'objc', 'objcpp' },
                      },
                  }
              })
```
I tried pasting it inside the cmd table but it did not seem to work, do you have any idea where it could go?

**Reply by HighCommander4** [reply]
The `--style` flag mentioned in my last comment is not something clangd currently supports, I was saying that would be a nice feature to add that would address this use case.

**Reply by HighCommander4** [reply]
A recent issue was filed for this enhancement: https://github.com/clangd/clangd/issues/2397

**Comment by leet0rz** [comment]
> The `--style` flag mentioned in my last comment is not something clangd currently supports, I was saying that would be a nice feature to add that would address this use case.

Ah, yes it would be. Would love to have that configurable. Thanks.

**Reply by HighCommander4** [reply]
I updated the discussion title to make it clearer and easier to find in case other people have feedback

**Comment by wangp-blog** [comment]
I have a terrible problem. vscode's default code formatter seems to conflict with clangd (I don't have clang-format installed)
I can't configure the code formatting properly now, and no matter how I set it, the indentation is always 2 instead of the expected 4.
Does clangd support configuration of formatting styles? If yes, what should I do in vscode?

**Reply by HighCommander4** [reply]
Clangd respects clang-format's options, so to configure an indent width of 4 you'd create a `.clang-format` file containing `IndentWidth: 4`.

**Reply by wangp-blog** [reply]
Thank you for your answer.
When I found out in the actual test that sometimes the configuration did not work, I initially thought that vscode's default formatter and clangd conflicted, but this was removed after the test. Another problem is that the.clang-format file should be placed in the "current directory or parent directory", where is the directory referring to clangd's plug-in directory in vscode or my actual project directory?

**Reply by HighCommander4** [reply]
The search starts in the directory containing the file being formatted, and then looks in ancestor directories. So generally you'd put it in the project directory to apply to all files in the project (but you could also e.g. have one in your home directory to apply a default set of format options to all projects in your home directory that don't have their own).

**Reply by wangp-blog** [reply]
> The search starts in the directory containing the file being formatted, and then looks in ancestor directories. So generally you'd put it in the project directory to apply to all files in the project (but you could also e.g. have one in your home directory to apply a default set of format options to all projects in your home directory that don't have their own).

Thank you. I have a few questions I'd like to ask, and if you can take the time to answer them, it would be very much appreciated. 1. Through my tests, I found that c/c++ plugin, clangd and clang-format all support code formatting. And the clang-format plugin seems to have stopped updating, which means I should give it up. 2. Unfortunately, if I uninstall the clang-format plugin, the other two seem to conflict and can't be formatted correctly. What should I do? 3. There is no style in the configuration item of clangd. Is file by default? Can I specify the location of file? Or go back to where you started and use the c/c++ plugin for formatting?

**Reply by HighCommander4** [reply]
What is "c/c plug-in"?

**Reply by wangp-blog** [reply]
> What is "c/c plug-in"?

I'm sorry, I have corrected it (c/c++ plugin).

**Reply by HighCommander4** [reply]
Do you mean https://marketplace.visualstudio.com/items?itemName=ms-vscode.cpptools?

Clangd is generally intended to be an alternative to Microsoft's plugin.

Can you say more about what sort of formatting conflict are you having?

**Reply by wangp-blog** [reply]
Yes, thank you for your answer. Clangd's formatting style is relatively stiff and does not support style, which may be the problem. My questions have been resolved and I am now using Microsoft Plug-ins, clangd, and clang-format, everything's fine.

**Reply by wangp-blog** [reply]
~I have a problem, I don't know if it is my setting error. I'm developing in c++, and when I output a semicolon or a newline, clang-format automatically indents 8 characters, but I set it to 4 characters. If I format the entire file or function, it is normal 4 characters at this point. I don't know how I should configure it, my Settings are: **"BasedOnStyle: Google, IndentWidth: 4, BreakBeforeBraces: Allman,IndentCaseLabels: false"**~

**Reply by HighCommander4** [reply]
@wangp-blog please file a new issue with details including [clangd logs](https://clangd.llvm.org/troubleshooting#gathering-logs).

**Reply by wangp-blog** [reply]
Thank you. I will edit it.

**Comment by leet0rz** [comment]
```lua
  lspconfig.lua_ls.setup({
    capabilities = capabilities,
    on_attach = on_attach,
    settings = {
      Lua = {
        runtime = {
          -- Tell the language server which version of Lua you're using (most likely LuaJIT in the case of Neovim)
          version = "LuaJIT",
        },
        diagnostics = {
          -- Get the language server to recognize the `vim` global
          globals = { "vim" },
        },
        workspace = {
          -- Make the server aware of Neovim runtime files
          library = {
            vim.api.nvim_get_runtime_file("", true),
            vim.fn.stdpath("config") .. "/lua",
          },
          checkThirdParty = false,
        },
        telemetry = { enable = false },
        format = {
          enable = true,
          defaultConfig = {
            indent_style = "space",
            indent_size = "2",
          }
        }
      },
    },
  })
  ```
  
  Here is an example of how lua_ls does it, was hoping clangd would be able to do the same without having to pollute the dir with another file.

**Comment by D029-W** [comment]
On top of that, I suggest allowing common custom style options for embedded `clang-format` in `clangd`.

In my workflow, I have several `.clang-format` files: one is for my personal preferences, used for reading and writing code, and others are kept in different project root directories for team style consistency. My editor is configured to invoke `clang-format` with the `--style=file:path/to/my/.clang-format` option upon opening or editing a file, reverting to the project’s format when I close it.

Currently, I have not found a way to directly pass this option to `clang-format` through `clangd`, and I have not been able to utilize the formatting feature embedded within `clangd`, as it only respects format files in the project directories.

I use the `--style=file:/path/to/my/.clang-format` option frequently, people may likely find other `clang-format` options useful too. And the impending “format-as-you-type” feature would be even more powerful if we could pass options to `clang-format`. 

Therefore, I suggest allowing common custom style options for `clang-format` in `clangd`.

**Comment by 424778940z** [comment]
run it to the same issue, do we have any progress regarding this? not be able to specify a custom file is really inconvenient....

**Reply by HighCommander4** [reply]
> do we have any progress regarding this?

You can follow https://github.com/clangd/clangd/issues/2397 for updates.

Note that there is a draft patch linked from that issue, but no one seems to be actively working on it. If you have the time and motivation to take it over and complete work on it, I think a number of users would appreciate it.
